design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+For pods that implement these probes, the upgrade controller can check them as another data point in determining if things are healthy before proceeding with the upgrade.
+
+### **Rollback**
+If the upgrade controller observes the cluster to be in an unhealthy state (that does not recover) during the upgrade process, it will need to roll back components in the cluster to the previous stable version.
+This is possible due to the rolling/canary approach of the upgrade controller.
+To roll a component back to the previous version, the controller will simply set the `image` field of the pod template spec to the previous version then terminate each pod to allow their managing controller to start a new pod on the *old* version to replace it.
+
+The hope is that cluster health and stability will be restored once it has been rolled back to the previous version, but it is possible that simply rolling back the version may not solve all cases of cluster instability that begin during an upgrade process.
+We will need more hands on experience with cluster upgrades in order to improve both upgrade reliability and rollback effectiveness.
+
+### **Experience**
+#### Automatic Upgrades
+As described earlier, the cluster administrator manually begins the upgrade process by editing the cluster CRD to indicate a new software version is desired.
+However, it is possible to also implement automatic upgrades if the user has opted in, perhaps by a new boolean field in the cluster CRD such as `autoUpgrade: true`.
+If this flag is set, the operator can periodically check Rook's container registry to see if there is a new version tag greater than the version tag currently deployed to the cluster.
+If a new version is found, then the upgrade controller will proceed to update the cluster to the new version in a rolling fashion as described earlier.
+
+It is also possible for the user to be using release channels instead of specific version numbers, e.g. `beta` or `stable`.
+In that case, the operator would be looking for a newer version that has been promoted to the channel that the user has specified.
+
+### `rookctl`
+We should consider implementing a few new commands in `rookctl` that will help the user monitor and verify the upgrade progress and status.
+Some examples for potential new commands would be:
+* `rookctl versions`: This command would return the version of all Rook components in the cluster, so they can see at a glance which components have finished upgrading.
+This is similar to the new [`ceph versions` command](http://ceph.com/community/new-luminous-upgrade-complete/).
+* `rookctl status --upgrade`:  This command would return a summary, retrieved from the upgrade controller, of the most recent completed steps and status of the upgrade that it is currently working on.
+
+### **Migrations and Breaking Changes**
```

Comment:
This got me thinking about setting `noout` during the OSD upgrade process. With `noout` set the cluster won't try to rebalance during OSD restart if for any reason it takes longer than usual. Although I don't recall seeing it used in the ceph upgrade docs, it may have some use in those cases of breaking changes or rollbacks to avoid the cluster going into a completely bad state.

Reply:
That's a great idea @hunter, I'll definitely incorporate `noout`.  I'll probably dig around a bit more too and see if there are other settings or commands we can use to better notify and prepare the Ceph layer that we're about to be recycling a lot of its instances...thanks!